// g2-for-react/lib/G2/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _g = require('@antv/g2');

var _g2 = _interopRequireDefault(_g);

var _fastDeepEqual = require('fast-deep-equal');

var _fastDeepEqual2 = _interopRequireDefault(_fastDeepEqual);

var _sizeSensor = require('size-sensor');

var _const = require('./const');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var G2 = function (_Component) {
  (0, _inherits3.default)(G2, _Component);

  function G2(props) {
    (0, _classCallCheck3.default)(this, G2);

    var _this = (0, _possibleConstructorReturn3.default)(this, (G2.__proto__ || (0, _getPrototypeOf2.default)(G2)).call(this, props));

    _this.dom = undefined;
    _this.g2 = undefined;
    return _this;
  }

  (0, _createClass3.default)(G2, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var _this2 = this;

      var _props = this.props,
          className = _props.className,
          creator = _props.creator,
          onReady = _props.onReady,
          data = _props.data,
          cfg = (0, _objectWithoutProperties3.default)(_props, ['className', 'creator', 'onReady', 'data']);


      this.g2 = new _g2.default.Chart((0, _extends3.default)({
        container: this.dom,
        data: data,
        height: this.dom.clientHeight
      }, cfg));

      // 创建过程
      creator(this.g2);

      this.g2.render();

      (0, _sizeSensor.bind)(this.dom, function (ele) {
        _this2.g2.changeHeight(ele.clientHeight);
      });

      // 成功回调
      onReady(this.g2);
    }
  }, {
    key: 'componentDidUpdate',
    value: function componentDidUpdate(prevProps, prevState) {
      // 数据变更
      if (!(0, _fastDeepEqual2.default)(prevProps.data, this.props.data)) {
        this.g2.changeData(this.props.data);
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      (0, _sizeSensor.clear)(this.dom);

      this.g2.destroy();
    }
  }, {
    key: 'render',
    value: function render() {
      var _this3 = this;

      var className = this.props.className;

      return _react2.default.createElement('div', {
        className: 'g2-for-react ' + (className || ''),
        style: { height: '100%', overflow: 'hidden' },
        ref: function ref(e) {
          _this3.dom = e;
        }
      });
    }
  }]);
  return G2;
}(_react.Component);

exports.default = G2;


G2.propTypes = {
  className: _propTypes2.default.string,
  data: _propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.any), _propTypes2.default.object]).isRequired,
  creator: _propTypes2.default.func,
  onReady: _propTypes2.default.func,
  padding: _propTypes2.default.any,
  background: _const.BGType,
  plotBackground: _const.BGType,
  forceFit: _propTypes2.default.bool,
  animate: _propTypes2.default.bool,
  pixelRatio: _propTypes2.default.number,
  theme: _propTypes2.default.oneOf([_propTypes2.default.string, _propTypes2.default.object]),
  renderer: _propTypes2.default.string
};

G2.defaultProps = {
  onReady: function onReady() {},
  creator: function creator() {},
  padding: undefined,
  background: undefined,
  plotBackground: undefined,
  forceFit: true,
  animate: true,
  pixelRatio: undefined,
  theme: undefined,
  renderer: undefined
};
